python/ray/autoscaler/gcp/config.py

```diff
@@ -103,52 +104,66 @@ def generate_rsa_key_pair():
     return public_key, pem
 
 
-def _create_crm(gcp_credentials):
+def _create_crm(gcp_credentials=None):
     return discovery.build(
         "cloudresourcemanager", "v1", credentials=gcp_credentials)
 
 
-def _create_iam(gcp_credentials):
+def _create_iam(gcp_credentials=None):
     return discovery.build("iam", "v1", credentials=gcp_credentials)
 
 
-def _create_compute(gcp_credentials):
+def _create_compute(gcp_credentials=None):
     return discovery.build("compute", "v1", credentials=gcp_credentials)
 
 
-def fetch_gcp_credentials_from_provider_config(provider_config):
+def construct_clients_from_provider_config(provider_config):
     """
     Attempt to fetch and parse the JSON GCP credentials from the provider
     config yaml file.
     """
-    service_account_info_string = provider_config.get("gcp_credentials")
-    if service_account_info_string is None:
-        logger.info("gcp_credentials not found in cluster yaml file. "
-                    "Falling back to GOOGLE_APPLICATION_CREDENTIALS "
-                    "environment variable.")
+    gcp_credentials = provider_config.get("gcp_credentials")
+    if gcp_credentials is None:
+        logger.debug("gcp_credentials not found in cluster yaml file. "
+                     "Falling back to GOOGLE_APPLICATION_CREDENTIALS "
+                     "environment variable.")
         # If gcp_credentials is None, then discovery.build will search for
         # credentials in the local environment.
-        return None
+        return _create_crm(), \
+            _create_iam(), \
+            _create_compute()
+
+    assert ("type" in gcp_credentials), \
+        "gcp_credentials cluster yaml field missing 'type' field."
+    assert ("credentials" in gcp_credentials), \
+        "gcp_credentials cluster yaml field missing 'credentials' field."
+
+    cred_type = gcp_credentials["type"]
+    credentials_field = gcp_credentials["credentials"]
+
+    if cred_type == "service_account":
+        # If parsing the gcp_credentials failed, then the user likely made a
+        # mistake in copying the credentials into the config yaml.
+        try:
+            service_account_info = json.loads(credentials_field)
+        except json.decoder.JSONDecodeError:
+            raise RuntimeError(
+                "gcp_credentials found in cluster yaml file but "
+                "formatted improperly.")
+        credentials = service_account.Credentials.from_service_account_info(
+            service_account_info)
+    else:
```

Comment:
How would you feel about making this explicitly `token`?

If we're going to have an unrestricted `else` it should throw an exception (and then we can remove the enum value checks from earlier).

Reply:
Sure that works. I think I already removed the enum checks though because the schema validator automatically enforces those value requirements (i.e. if you put in something other than "service_account" or "credentials_token" it will tell you that that's not allowed)